ts/private/ts_project_worker.js

```diff
@@ -0,0 +1,384 @@
+}
+
+
+
+function isAncestorDirectory(ancestor, child) {
+    const ancestorChunks = ancestor.split(path.sep).filter(i => !!i);
+    const childChunks = child.split(path.sep).filter(i => !!i);
+    return ancestorChunks.every((chunk, i) => childChunks[i] === chunk)
+}
+
+/** @type {Map<string, {program: ts.EmitAndSemanticDiagnosticsBuilderProgram, host: ts.WatchCompilerHost, previousInputs?: import("@bazel/worker").Inputs}>} */
+const workers = new Map(); 
+
+function getTsConfigPath(args) {
+    return args[args.indexOf('--project') + 1]
+}
+
+function createProgram(args, initialInputs) {
+    const tsconfig = getTsConfigPath(args);
+    const execRoot = path.resolve(process.cwd(), '..', '..', '..')
+
+    debuglog(`tsconfig: ${tsconfig}`);
+    debuglog(`execroot ${execRoot}`);
+
+    const parsedArgs = ts.parseCommandLine(args)
+
+    const directoryWatchers = new Map();
+    const fileWatchers = new Map();
+
+    const knownInputs = new Set(initialInputs);
+
+    /** @type {ts.System} */
+    const sys = {
+        ...ts.sys,
+        write: worker.log,
+        writeOutputIsTTY: false,
+        setTimeout: setImmediate,
+        clearTimeout: clearImmediate,
+        fileExists: fileExists,
+        readFile: readFile,
+        writeFile: writeFile,
+        readDirectory: readDirectory,
+        directoryExists: directoryExists,
+        getDirectories: getDirectories,
+        watchFile: watchFile,
+        watchDirectory: watchDirectory
+    }
+
+    const host = ts.createWatchCompilerHost(
+        tsconfig,
+        parsedArgs.options,
+        sys,
+        createEmitCacheAndDiagnosticsProgram,
+        noop,
+        noop
+    )
+
+    host.invalidate = invalidate;
+
+    program = ts.createWatchProgram(host)
+    
+    function getDirectoryWatcherForPath(path) {
+        for (const [directory, cb] of directoryWatchers.entries()) {
+            if (isAncestorDirectory(directory, path)) {
+                debuglog(`found a directory watcher for ${path} ${directory}`)
+                return cb;
+            }
+        }
+    }
+
+    function writeFile(path, data, writeByteOrderMark) {
+        debuglog(`writeFile ${path}`);
+        ts.sys.writeFile(path, data, writeByteOrderMark);
+    }
+
+    function readFile(filePath, encoding) {
+        debuglog(`readFile ${filePath}`);
+        const relative = path.relative(execRoot, filePath)
+        /** if it is under node_modules just allow file reads as we don't have a list of deps */
+        if (filePath.indexOf("node_modules") == -1 && !knownInputs.has(relative)) {
+            return undefined
+        }
+        return ts.sys.readFile(filePath, encoding);
+    }
+
+    function directoryExists(directory) {
+        if (!directory.startsWith(execRoot)) {
+            return false;
+        }
+        const exists = ts.sys.directoryExists(directory)
+        directory.indexOf("node_modules") == -1 && debuglog(`directoryExists ${directory} ${exists}`);
```

Comment:
`if` to avoid this long line? `.includes` instead of `indexOf`...

Reply:
Agreed.